fuel-vm/src/constraints/reg_key.rs

```diff
@@ -0,0 +1,342 @@
+//! Utilities for accessing register values and proving at compile time that
+//! the register index is valid.
+//!
+//! This module also provides utilities for mutably accessing multiple registers.
+use std::ops::Deref;
+use std::ops::DerefMut;
+
+use fuel_asm::RegId;
+use fuel_asm::Word;
+
+use crate::consts::VM_REGISTER_COUNT;
+use crate::consts::VM_REGISTER_WRITE_COUNT;
+
+#[derive(Debug, PartialEq, Eq)]
+/// Mutable reference to a register value at a given index.
+pub struct RegMut<'r, const INDEX: u8>(&'r mut Word);
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+/// Immutable reference to a register value at a given index.
+pub struct Reg<'r, const INDEX: u8>(&'r Word);
+
+impl<'r, const INDEX: u8> RegMut<'r, INDEX> {
+    /// Create a new mutable register reference.
+    pub fn new(reg: &'r mut Word) -> Self {
+        Self(reg)
+    }
+}
+
+impl<'r, const INDEX: u8> Reg<'r, INDEX> {
+    /// Create a new immutable register reference.
+    pub fn new(reg: &'r Word) -> Self {
+        Self(reg)
+    }
+}
+
+impl<const INDEX: u8> Deref for Reg<'_, INDEX> {
+    type Target = Word;
+    fn deref(&self) -> &Self::Target {
+        self.0
+    }
+}
+impl<const INDEX: u8> Deref for RegMut<'_, INDEX> {
+    type Target = Word;
+    fn deref(&self) -> &Self::Target {
+        self.0
+    }
+}
+impl<const INDEX: u8> DerefMut for RegMut<'_, INDEX> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.0
+    }
+}
+impl<'a, const INDEX: u8> From<RegMut<'a, INDEX>> for Reg<'a, INDEX> {
+    fn from(reg: RegMut<'a, INDEX>) -> Self {
+        Self(reg.0)
+    }
+}
+impl<'r, const INDEX: u8> RegMut<'r, INDEX> {
+    /// Re-borrow the register as an immutable reference.
+    pub fn as_ref(&self) -> Reg<INDEX> {
+        Reg(self.0)
+    }
+}
+
+impl<'r, const INDEX: u8> RegMut<'r, INDEX> {
+    /// Re-borrow the register as a mutable reference.
+    pub fn as_mut(&mut self) -> RegMut<INDEX> {
+        RegMut(self.0)
+    }
+}
+
+macro_rules! impl_keys {
+    ( $($i:ident, $f:ident)* ) => {
+        $(
+            #[doc = "Register index key for use with Reg and RegMut."]
+            pub const $i: u8 = RegId::$i.to_u8();
+        )*
+        #[doc = "Get register reference by name."]
+        pub trait GetReg {
+        $(
+            #[doc = "Get register reference for this key."]
+            fn $f(&self) -> Reg<'_, $i>;
+        )*
+        }
+        impl GetReg for [Word; VM_REGISTER_COUNT] {
+        $(
+            fn $f(&self) -> Reg<'_, $i> {
+                Reg(&self[$i as usize])
+            }
+        )*
+        }
+    };
+}
+
+impl_keys! {
+    ZERO, zero
+    ONE, one
+    OF, of
+    PC, pc
+    SSP, ssp
+    SP, sp
+    FP, fp
+    HP, hp
+    ERR, err
+    GGAS, ggas
+    CGAS, cgas
+    BAL, bal
+    IS, is
+    RET, ret
+    RETL, retl
+    FLAG, flag
+}
+
+pub(crate) struct ReadRegisters<'a> {
```

Comment:
I think the Read vs Write distinction is somewhat confusing, since we need to modify these registers internally within the VM. Perhaps this should be renamed to something like `InternalRegisters` or `ProtectedRegisters`.